File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        
        mapp_closing = {
            
            ")" :"(",
            "}" : "{",
            "]" : "["
        }
        
        stack = []
        
        for i in s:
            
            if i in mapp_closing:
                
                # No matching opening bracket found expression is invalid.
                if not stack or stack[-1] != mapp_closing[i]:
                    
                    return False
                
                # Pop the opening bracket
                stack.pop()
                
            else:
                stack.append(i)
                
        return len(stack) == 0
                
                
                

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        
        Stack<Character> stack = new Stack<>();
        
        Map<Character,Character> map = new HashMap<>(3);
        
        map.put(')','(');
        map.put(']','[');
        map.put('}','{');
        
        for(char ch : s.toCharArray()){
            if(ch== '(' || ch =='[' || ch =='{') stack.add(ch);
            else if( !stack.isEmpty() && stack.peek() == map.get(ch)) 
                stack.pop();
            
            else
                return false;
            
        }
    
        return stack.empty();
    }
}